### PR TITLE
Might be null

### PR DESCRIPTION
**Pull request by manuelsblanco**
**This PR refactors the VmMap class to enhance immutability handling and improve overall performance.** 

**Key changes include:**

- Refactored the VmMap class to ensure consistent use of TruffleBoundary annotations.
- Improved the Builder class for more efficient map construction.
- Added null checks and robustness improvements for handling empty collections.
- Optimized methods related to map and key order immutability.
- These changes aim to increase the stability and reliability of the VM runtime. Please review the updates, and let me know if any adjustments are needed.

### CODE REVIEW

File: bench/src/jmh/java/org/pkl/core/Fibonacci.java [hunk 1]
```diff
@@ -98,7 +100,23 @@ public long fib_lambda() {
...
+
+    // If the result for n is already calculated, return it from memo
+    if (memo.containsKey(n)) {
+      return memo.get(n);
+    }
+
+    // Otherwise, calculate Fibonacci recursively and store the result in memo
+    long result = fib(n - 1) + fib(n - 2);
+    memo.put(n, result);
+    return result;
```
Comment (bioball) [inline bench/src/jmh/java/org/pkl/core/Fibonacci.java:120]: This class is intentionally written this way--we use this to compare Pkl's fibonacci performance to native Java (see comment on line 99)

File: pkl-core/src/main/java/org/pkl/core/runtime/VmMap.java [hunk 2]
```diff
@@ -98,7 +99,7 @@ public boolean hasNext() {
       @TruffleBoundary
       public Map.Entry<Object, Object> next() {
         var key = keyIterator.next();
-        return Map.entry(key, map.get(key));
+        return Map.entry(key, Objects.requireNonNull(map.get(key)));
```
Comment (bioball) [inline pkl-core/src/main/java/org/pkl/core/runtime/VmMap.java:102]: Mmm I'm not sure why GitHub's diff is showing this, but this is out of date with what's on our main branch.

We already have a non-null assertion here, see:

https://github.com/apple/pkl/blob/a9c4963a4a60e06af683f4a700f290b75ca99d22/pkl-core/src/main/java/org/pkl/core/runtime/VmMap.java#L101-L103

**Review by bioball — COMMENTED** [review]
Thanks for the PR! Unfortunately, I don't think there's anything here that we can accept for contribution (see my comments)

**Comment by manuelsblanco** [comment]
Thanks for the reply.

**Pull request closed without merging**